[AKS] `az aks nodepool delete`: Add `--ignore-pod-disruption-budget` option for ignoring PodDisruptionBudget

tonychen15 opened the pull request:
**Related command**
az aks nodepool delete

**Description**<!--Mandatory-->
Add `--ignore-pod-disruption-budget` option for ignoring PodDisruptionBudget. This will expedite the nodepool deletion which may be delayed or blocked by PDB.

**Testing Guide**
Use the command `az aks nodepool delete --ignore-pod-disruption-budget` to send the request to aks. 

**History Notes**
[AKS] `az aks nodepool delete`: Add `--ignore-pod-disruption-budget` option for ignoring PodDisruptionBudget

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

yonzhan commented:
Thank you for your contribution! We will review the pull request and get back to you soon.

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py:
```diff
@@ -1504,6 +1504,19 @@ def get_disable_windows_outbound_nat(self) -> bool:
         """
         return self._get_disable_windows_outbound_nat()
 
+    def get_ignore_pod_disruption_budget(self) -> bool:
+        return self._get_ignore_pod_disruption_budget()
+
+    def _get_ignore_pod_disruption_budget(self) -> bool:
+        """Obtain the value of ignore_pod_disruption_budget, default value is False.
+
+        :return: bool
+        """
+        # read the original value passed by the command
+        ignore_pod_disruption_budget = self.raw_param.get("ignore_pod_disruption_budget", False)
+
+        # This parameter does not need dynamic completion.
+        return ignore_pod_disruption_budget
```
I think you don't need this as this is not an option for `az aks nodepool add/update`

tonychen15 replied:
I see. I overlooked your comment in the code. Removed.

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_agentpool_decorator.py:
```diff
@@ -1494,6 +1494,18 @@ def common_get_drain_timeout(self):
         ctx_1.attach_agentpool(agentpool_1)
         self.assertEqual(ctx_1.get_drain_timeout(), 123)
 
+    def common_get_ignore_pod_disruption_budget(self):
+        ctx_1 = AKSAgentPoolContext(
+            self.cmd,
+            AKSAgentPoolParamDict({
+                "ignore_pod_disruption_budget": True,
+            }),
+            self.models,
+            DecoratorMode.DELETE,
+            self.agentpool_decorator_mode,
+        )
+        self.assertEqual(ctx_1.get_ignore_pod_disruption_budget(), True)
+
```
And there's no need for the test here.

tonychen15 replied:
Removed, thanks.

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/_params.py:
```diff
@@ -929,6 +929,14 @@ def load_arguments(self, _):
+    with self.argument_context("aks nodepool delete") as c:
+        c.argument(
+            "ignore_pod_disruption_budget",
+            options_list=["--ignore-pod-disruption-budget"],
```
If the only option form is `--ignore-pod-disruption-budget`, you could ignore this line. In cli-extensions/aks-preview, it has another shorthand option name `-i`, do you want to keep it?

tonychen15 replied:
Got it. I intentionally remove that to make sure CX fully understand what they're doing.

tonychen15 replied:
Need to add ignore-pdb as ignore-pod-disruption-budget exceed the maximum length 22.

FumingZhang reviewed (commented):
Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=106793527&view=results) to validate the change.

tonychen15 commented:
It seems both Python39 and Python312 failed at test case aks_agentpool_create_scale_delete with version 2020-09-01 (https://dev.azure.com/azclitools/public/_build/results?buildId=201476&view=logs&j=4e4d232d-230a-55c9-0730-4c1a9c36b9e5&t=6d1bdb99-245b-5117-da18-01d8df6f72c0&l=1356). Is this failed test case a known issue?

tonychen15 commented:
What's the best practice for the option length is too long issue?

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/custom.py:
```diff
@@ -2610,7 +2610,7 @@ def aks_agentpool_delete(cmd, client, resource_group_name, cluster_name,
                        "use 'aks nodepool list' to get current node pool list".format(nodepool_name))
 
-    return sdk_no_wait(no_wait, client.begin_delete, resource_group_name, cluster_name, nodepool_name)
+    return sdk_no_wait(no_wait, client.begin_delete, resource_group_name, cluster_name, nodepool_name, ignore_pod_disruption_budget=ignore_pod_disruption_budget)
```
```suggestion
    from azure.cli.core.cloud import get_active_cloud
    active_cloud = get_active_cloud(cmd.cli_ctx)
    if active_cloud.profile != "latest":
        return sdk_no_wait(
            no_wait,
            client.begin_delete,
            resource_group_name,
            cluster_name,
            nodepool_name,
        )
    else:
        return sdk_no_wait(
            no_wait,
            client.begin_delete,
            resource_group_name,
            cluster_name,
            nodepool_name,
            ignore_pod_disruption_budget=ignore_pod_disruption_budget,
        )
```
something like this can be used as a workaround to support both the old profile with 2020-11-01 as the default API version and the latest profile with the latest API version.

Please move `from azure.cli.core.cloud import get_active_cloud` to the top of this file

tonychen15 replied:
Added. It seems it still doesn't work.

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/_params.py:
```diff
@@ -929,6 +929,14 @@ def load_arguments(self, _):
+    with self.argument_context("aks nodepool delete") as c:
+        c.argument(
+            "ignore_pod_disruption_budget",
+            options_list=["--ignore-pod-disruption-budget, --ignore-pdb"],
```
You can add a rule exclusion to `src/azure-cli/azure/cli/command_modules/acs/linter_exclusions.yml`

tonychen15 replied:
Added.

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/_params.py:
```diff
@@ -919,6 +919,14 @@ def load_arguments(self, _):
+    with self.argument_context("aks nodepool delete") as c:
+        c.argument(
+            "ignore_pod_disruption_budget",
+            options_list=["--ignore-pdb"],
```
```suggestion
            options_list=["--ignore-pdb", "--ignore-pod-disruption-budget"],
```
please also include the default option name

tonychen15 replied:
Added back, hope that pylint not to report length too long error.

FumingZhang commented:
Requeued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=107858522&view=results).

The following cases failed in replay mode as you've changed the behavior of `az aks nodepool delete`.

- test_aks_create_with_windows_gmsa
- test_aks_create_with_windows_msi
- test_aks_nodepool_create_scale_delete_msi
- test_aks_nodepool_delete_with_ignore_pod_disruption_budget
- test_aks_nodepool_system_pool_msi
- test_aks_update_with_windows_gmsa
- test_aks_update_with_windows_password

There's a PR #30290 that bumped the default API version merged earlier today. Please rebase from dev branch, requeue the tests and them commit the updated recording files.

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/_help.py:
```diff
@@ -1614,6 +1614,14 @@
 type: command
 short-summary: Delete the agent pool in the managed Kubernetes cluster.
+parameters:
+    - name: --ignore-pdb
```
> -  FAIL - HIGH severity: unrecognized_help_parameter_rule
>     Help-Entry: `aks nodepool delete` - The following parameter help names are invalid: --ignore-pdb

Try the following, one of them would be correct
```suggestion
    - name: --ignore-pdb --ignore-pod-disruption-budget
```

```suggestion
    - name: --ignore-pod-disruption-budget --ignore-pdb
```

tonychen15 replied:
Added.

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/acs/_help.py:
```diff
@@ -1614,6 +1614,14 @@
+parameters:
+    - name: --ignore-pod-disruption-budget --ignore-pdb
+      type: bool
+      short-summary: ignore-pdb deletes an existing nodepool without considering Pod Disruption Budget.
```
```suggestion
      short-summary: Delete an existing node pool without considering Pod Disruption Budget.
```

zhoxing-ms commented on src/azure-cli/azure/cli/command_modules/acs/custom.py:
```diff
@@ -2604,7 +2605,23 @@ def aks_agentpool_delete(cmd, client, resource_group_name, cluster_name,
                        "use 'aks nodepool list' to get current node pool list".format(nodepool_name))
 
-    return sdk_no_wait(no_wait, client.begin_delete, resource_group_name, cluster_name, nodepool_name)
+    active_cloud = get_active_cloud(cmd.cli_ctx)
+    if active_cloud.profile != "latest":
```
Will this logic also be used in old profiles env such as Azure Stack?

tonychen15 replied:
No, this feature is only used in AKS. @zhoxing-ms

tonychen15 commented:
@FumingZhang and @zhoxing-ms I have rebased to the latest dev branch, and updated per your comments, somehow that full test is still failed. As ignore-pdb is not a required parameter to "az aks nodepool delete", so it should not have the impact to other aks cli command. I'm quite confused about why those aks tests failed for it. 

Any more suggestions on this? Thanks.

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/_help.py:
```diff
@@ -1614,6 +1614,14 @@
 type: command
 short-summary: Delete the agent pool in the managed Kubernetes cluster.
+parameters:
+    - name: --ignore-pod-disruption-budget --ignore-pdb
```
```suggestion
    - name: --ignore-pdb --ignore-pod-disruption-budget
```

tonychen15 replied:
Yeah, it works now. Thanks a lot. 

I don't expect it is so strict for a parameter order in the helper file.

FumingZhang commented:
Failures in Automation Full Test XXX are not related to your changes, I think the corresponding team is working on the fix. As for failures in linter, may try the other way https://github.com/Azure/azure-cli/pull/30196/files#r1837556755

> -  FAIL - HIGH severity: unrecognized_help_parameter_rule
>     Help-Entry: `aks nodepool delete` - The following parameter help names are invalid: --ignore-pod-disruption-budget --ignore-pdb

tonychen15 commented:
> Failures in Automation Full Test XXX are not related to your changes, I think the corresponding team is working on the fix. As for failures in linter, may try the other way https://github.com/Azure/azure-cli/pull/30196/files#r1837556755
> 
> > * FAIL - HIGH severity: unrecognized_help_parameter_rule
> >   Help-Entry: `aks nodepool delete` - The following parameter help names are invalid: --ignore-pod-disruption-budget --ignore-pdb

So, we will wait for another team's fix, then merge this PR? If so, I will retest it later.

tonychen15 commented:
> > Failures in Automation Full Test XXX are not related to your changes, I think the corresponding team is working on the fix. As for failures in linter, may try the other way https://github.com/Azure/azure-cli/pull/30196/files#r1837556755
> > > * FAIL - HIGH severity: unrecognized_help_parameter_rule
> > >   Help-Entry: `aks nodepool delete` - The following parameter help names are invalid: --ignore-pod-disruption-budget --ignore-pdb
> 
> So, we will wait for another team's fix, then merge this PR? If so, I will retest it later.

@FumingZhang @zhoxing-ms any way we can move this PR forward?

FumingZhang commented:
@tonychen15 please fix the failed CI check.

Take [Azure.azure-cli Full Test (Automation Full Test Python312 Profile Latest instance6)](https://dev.azure.com/azclitools/public/_build/results?buildId=205542&view=logs&j=950e443e-fa90-5f50-4303-cd9168964025&t=e8e1e5d0-e808-59fd-7db9-daddfd146787) as an example, the following cases failed

> FAILED src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py::AzureKubernetesServiceScenarioTest::test_aks_create_with_windows_gmsa
> FAILED src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py::AzureKubernetesServiceScenarioTest::test_aks_create_with_windows_msi
> FAILED src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py::AzureKubernetesServiceScenarioTest::test_aks_nodepool_create_scale_delete_msi
> FAILED src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py::AzureKubernetesServiceScenarioTest::test_aks_nodepool_system_pool_msi
> FAILED src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py::AzureKubernetesServiceScenarioTest::test_aks_update_with_windows_gmsa
> FAILED src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py::AzureKubernetesServiceScenarioTest::test_aks_update_with_windows_password

and the error is

> >           raise AssertionError(ex)
> E           AssertionError: Can't overwrite existing cassette ('/mnt/vss/_work/1/s/src/azure-cli/azure/cli/command_modules/acs/tests/latest/recordings/test_aks_update_with_windows_password.yaml') in your current record mode ('once').
> E           No match for the request (<Request (DELETE) [https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest000001/providers/Microsoft.ContainerService/managedClusters/cliakstest000001/agentPools/npwin?api-version=2024-09-01&ignore-pod-disruption-budget=false>](https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest000001/providers/Microsoft.ContainerService/managedClusters/cliakstest000001/agentPools/npwin?api-version=2024-09-01&ignore-pod-disruption-budget=false%3E)) was found.
> E           Found 1 similar requests with 1 different matcher(s) :
> E           
> E           1 - (<Request (DELETE) [https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest000001/providers/Microsoft.ContainerService/managedClusters/cliakstest000001/agentPools/npwin?api-version=2024-09-01>).](https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/clitest000001/providers/Microsoft.ContainerService/managedClusters/cliakstest000001/agentPools/npwin?api-version=2024-09-01%3E).)
> E           Matchers succeeded : ['method', 'scheme', 'host', 'port', 'path']
> E           Matchers failed :
> E           _custom_request_query_matcher - assertion failure :
> E           None

As I've mentioned
> Requeued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=107858522&view=results).
> 
> **The following cases failed in replay mode as you've changed the behavior of `az aks nodepool delete`.**
> 
> * test_aks_create_with_windows_gmsa
> * test_aks_create_with_windows_msi
> * test_aks_nodepool_create_scale_delete_msi
> * test_aks_nodepool_delete_with_ignore_pod_disruption_budget
> * test_aks_nodepool_system_pool_msi
> * test_aks_update_with_windows_gmsa
> * test_aks_update_with_windows_password
> 
> There's a PR #30290 that bumped the default API version merged earlier today. Please rebase from dev branch, requeue the tests and them commit the updated recording files.

I just requeued the [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=108602025&view=results) for you. Please commit the updated recording files once those tests passed.

tonychen15 commented:
I see. I need to manually rerun those failed test cases locally and upload their logs. I thought the pipeline may rerun them on the server side.
test_aks_create_with_windows_gmsa
test_aks_create_with_windows_msi
test_aks_nodepool_create_scale_delete_msi
test_aks_nodepool_delete_with_ignore_pod_disruption_budget
test_aks_nodepool_system_pool_msi
test_aks_update_with_windows_gmsa
test_aks_update_with_windows_password

FumingZhang commented:
I noticed that you did not upload the new recording files corresponding to the above-mentioned test cases. The live test I queued passed, and you could find them from [pipeline artifacts](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=108602025&view=artifacts&pathAsName=false&type=publishedArtifacts).

<img width="760" alt="Screenshot 2024-12-03 144445" src="https://github.com/user-attachments/assets/5911b2e4-a7f9-467a-a320-1f895ee87eab">

FumingZhang commented:
/azp run

yanzhudd commented:
Please resolve the CI issues.

yanzhudd commented on src/azure-cli/azure/cli/command_modules/acs/_help.py:
```diff
@@ -1668,9 +1668,13 @@
     short-summary: The value provided will be compared to the ETag of the node pool, if it matches the operation will proceed. If it does not match, the request will be rejected to prevent accidental overwrites. This must not be specified when creating a new agentpool.
+  - name: --ignore-pdb
+    type: bool
+    short-summary: ignore-pdb deletes an existing nodepool without considering Pod Disruption Budget.
```
```suggestion
    short-summary: Delete an existing nodepool without considering Pod Disruption Budget.
```

tonychen15 replied:
Updated.

yanzhudd commented on src/azure-cli/azure/cli/command_modules/acs/_params.py:
```diff
@@ -949,6 +949,14 @@ def load_arguments(self, _):
+            "ignore_pdb",
+            options_list=["--ignore-pdb", "--ignore-pod-disruption-budget"],
+            action='store_true',
+            help="delete an AKS nodepool by ignoring PodDisruptionBudget setting",
```
```suggestion
            help="Delete an AKS nodepool by ignoring PodDisruptionBudget setting",
```

tonychen15 replied:
Changed.

yanzhudd commented:
Please note Azure CLI will freeze the code on 01/28/2025 10:00 UTC for the upcoming release. If you want to catch this release train, please resolve the comments ASAP, otherwise it has to be postponed to next sprint.

yanzhudd commented:
please fix the CI issues.

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/agentpool_decorator.py:
```diff
@@ -1519,6 +1519,20 @@ def get_if_none_match(self) -> str:
         """
         return self.raw_param.get("if_none_match")
 
+    def get_ignore_pod_disruption_budget(self) -> bool:
```
Where is the value of this option passed?

FumingZhang replied:
Seems the option is only used by command `aks nodepool delete`, if that's the case, you could undo your changes in this file as the codes here are only for `aks nodepool add/update` and `aks create/update`.

tonychen15 commented:
I don't know how this help name --ignore-pdb is invalid. it is accepted by the _params.py, why it is not accepted by _help.py? whether they have different validation criteria? The failed two checks failed at the same place. @yanzhudd @FumingZhang 

"
FAIL - HIGH severity: unrecognized_help_parameter_rule
    Help-Entry: `aks nodepool delete` - The following parameter help names are invalid: --ignore-pdb
"

FumingZhang commented on src/azure-cli/azure/cli/command_modules/acs/_help.py:
```diff
@@ -1668,9 +1668,12 @@
   - name: --if-match
     type: string
     short-summary: The value provided will be compared to the ETag of the node pool, if it matches the operation will proceed. If it does not match, the request will be rejected to prevent accidental overwrites. This must not be specified when creating a new agentpool.
+  - name: --ignore-pdb
```
the linter CI failed as you added another option form (`-i`) for this parameter, so you'll need to change this help info to one of the followings (I'm not sure which order the CI rule prefers, but it only considers one to be correct)
```suggestion
  - name: --ignore-pdb -i
```
```suggestion
  - name: -i --ignore-pdb
```

tonychen15 replied:
Thanks for the suggestion. Hope it will work. 
If it works, the git warning is somehow misleading and it should indicate one option is missed but not ignore-pdb is invalid.

FumingZhang reviewed (approved):
LGTM

Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=115576746&view=results) to validate the change, test passed!

tonychen15 commented:
@jsntcy and @yonzhan could you take a look this test ready PR? Thanks.
